_posts/2020-08-02-prediction-of-medical-costs/2020-08-02-prediction-of-medical-costs.Rmd

```diff
@@ -0,0 +1,600 @@
+  hp_ranger$min.node.size[i] = ranger[[i]]$tuning_result$params$min.node.size
+}
+
+
+plot_hp_ranger_1 = ggplot(data = hp_ranger) +
+  geom_boxplot(aes(x = 1, y = unlist(min.node.size))) +
+  ylim(c(1, 50)) +
+  geom_hline(yintercept = 1, col = "red", linetype = "dashed") +
+  geom_hline(yintercept = 50, col = "red", linetype = "dashed") +
+  labs(y = "min.node.size", x = "", subtitle = "Learner: Random Forest \nHP: min.node.size") +
+  theme_bw() +
+  theme(axis.text.x = element_blank())
+
+plot_hp_ranger_2 = ggplot(data = hp_ranger) +
+  geom_boxplot(aes(x = 1, y = unlist(mtry))) +
+  ylim(c(2, 6)) +
+  geom_hline(yintercept = 2, col = "red", linetype = "dashed") +
+  geom_hline(yintercept = 6, col = "red", linetype = "dashed") +
+  labs(y = "mtry", x = "", subtitle = "Learner: Random Forest \nHP: mtry") +
+  theme_bw() +
+  theme(axis.text.x = element_blank())
+
+png("hyperparam_setting.png")
+grid.arrange(plot_hp_knn, plot_hp_rpart_1, plot_hp_rpart_2, plot_hp_ranger_1, plot_hp_ranger_2,
+  nrow = 2,
+  top = textGrob("Tuning Results: Hyperparameter Settings", gp = gpar(fontface = 2, fontsize = 14)))
+dev.off()
+```
+
+![Tuning Results: Hyperparameter Setting](hyperparam_setting.png)
+
+These plots show the range of optimal hyperparameters obtained by the inner resampling loop.
+So, each boxplot is based on 5 values. The red lines (dashed) indicate the range, within which the hyperparameters were optimized.
+It can, for example, be seen that while various different $k$-values are deemed optimal for kNN, $mtry = 4$ is determined the optimal setting for the Ranger algorithm in every iteration.
+Also, the optimal $cp$ in CART seems to lie within a small range, while the best-performing $minsplit$-parameter is rather volatile.
+However, the illustration neglects the fact that in tuning we are looking at combinations of hyperparameters and the more dimensions we consider, the fewer of the hyperparameter space we are investigating with a fixed number of evaluations in the tuning.    
+
+### Outer Loop: Model Selection
+
+For nested resampling, we also define a 5-fold cross validation as our  outer loop.
+This means, we obtain five pairs of training and test sets. Each of those initiates the tuning process for the hyperparameter spaces specified earlier. 
+For each of the five iterations, the learner is fitted on the outer training set while its performance is evaluated on the outer test set.
+By averaging these five performance estimates of the outer loops, we get an unbiased performance estimate for the learner.  
+
+As mentioned earlier, we aim to compare the performance of different learning algorithms on the prediction of insurance costs.
+By initiating a benchmarking process, we can combine all candidate models in one single nested resampling process.
+All we have to do is to define a grid with the regression task, the (tuned) learners and the outer resampling strategy.  
+
+```{r benchmark, echo=TRUE, eval=longrun}
```

Comment:
This is the same as above?

Reply:
jap but in the first time the chunk wont be shown but evaluated, to have access to the autotuners...
In the second chapter "outer loop" this chunk is shown but not evaluated to vizualize that benchmarking is done now....
Guess the text and order makes sense once you knit it...